zcash_primitives/src/transaction/fees/zip317.rs

```diff
@@ -0,0 +1,101 @@
+};
+
+/// A [`FeeRule`] implementation that implements the [ZIP 317] fee rule.
+///
+/// [`FeeRule`]: crate::transaction::fees::FeeRule
+/// [ZIP 317]: https//zips.z.cash/zip-0317
+#[derive(Clone, Debug)]
+pub struct FeeRule {
+    marginal_fee: Amount,
+    grace_actions: usize,
+    p2pkh_standard_input_size: usize,
+    p2pkh_standard_output_size: usize,
+}
+
+impl FeeRule {
+    /// Construct a new FeeRule instance with the specified parameter values.
+    /// Returns `None` if either `p2pkh_standard_input_size` or `p2pkh_standard_output_size`
+    /// are zero.
+    pub fn new(
+        marginal_fee: Amount,
+        grace_actions: usize,
+        p2pkh_standard_input_size: usize,
+        p2pkh_standard_output_size: usize,
+    ) -> Option<Self> {
+        if p2pkh_standard_input_size == 0 || p2pkh_standard_output_size == 0 {
+            None
+        } else {
+            Some(Self {
+                marginal_fee,
+                grace_actions,
+                p2pkh_standard_input_size,
+                p2pkh_standard_output_size,
+            })
+        }
+    }
+
+    /// Construct a new FeeRule using the standard ZIP 317 constants.
+    pub fn standard() -> Self {
+        Self {
+            marginal_fee: Amount::from_u64(5000).unwrap(),
+            grace_actions: 2,
+            p2pkh_standard_input_size: 150,
+            p2pkh_standard_output_size: 34,
+        }
+    }
+
+    /// Returns the ZIP 317 marginal fee.
+    pub fn marginal_fee(&self) -> Amount {
+        self.marginal_fee
+    }
+    /// Returns the ZIP 317 number of grace actions
+    pub fn grace_actions(&self) -> usize {
+        self.grace_actions
+    }
+    /// Returns the ZIP 317 standard P2PKH input size
+    pub fn p2pkh_standard_input_size(&self) -> usize {
+        self.p2pkh_standard_input_size
+    }
+    /// Returns the ZIP 317 standard P2PKH output size
+    pub fn p2pkh_standard_output_size(&self) -> usize {
+        self.p2pkh_standard_output_size
+    }
+}
+
+impl super::FeeRule for FeeRule {
+    type Error = BalanceError;
+
+    fn fee_required<P: consensus::Parameters>(
+        &self,
+        _params: &P,
+        _target_height: BlockHeight,
+        transparent_inputs: &[impl transparent::InputView],
+        transparent_outputs: &[impl transparent::OutputView],
+        sapling_input_count: usize,
+        sapling_output_count: usize,
+    ) -> Result<Amount, Self::Error> {
+        let t_in_total_size = transparent_inputs.len() * 150;
+        let t_out_total_size = transparent_outputs.len() * 34;
+
+        let ceildiv = |num: usize, den: usize| (num / den) + if num % den == 0 { 0 } else { 1 };
+
+        let logical_actions = max(
+            ceildiv(t_in_total_size, self.p2pkh_standard_input_size),
+            ceildiv(t_out_total_size, self.p2pkh_standard_output_size),
+        ) + max(sapling_input_count, sapling_output_count);
+
+        (self.marginal_fee * max(logical_actions, self.grace_actions)).ok_or(BalanceError::Overflow)
```

Comment:
Nit: to more closely match ZIP 317:
```suggestion
        (self.marginal_fee * max(self.grace_actions, logical_actions)).ok_or(BalanceError::Overflow)
```